src/pytesmo/validation_framework/data_averaging.py

```diff
@@ -0,0 +1,428 @@
+            if e.args[0] == "No such file or directory":
+                warnings.warn(
+                    "IOError while reading dataset {} with args {:}".format(name,
+                                                                            args))
+                return None
+            else:
+                raise e
+
+        if len(data_df) == 0:
+            warnings.warn("No data for dataset {}".format(name))
+            return None
+
+        if not isinstance(data_df, pd.DataFrame):
+            warnings.warn("Data is not a DataFrame {:}".format(args))
+            return None
+
+        if self.period is not None:
+            # here we use the isoformat since pandas slice behavior is
+            # different when using datetime objects.
+            data_df = data_df[
+                      self.period[0].isoformat():self.period[1].isoformat()]
+
+        if len(data_df) == 0:
+            warnings.warn("No data for dataset {} with arguments {:}".format(name,
+                                                                             args))
+            return None
+
+        else:
+            return data_df
+
+
+class DataAverager(MixinReadTs):
+    """
+    This class provides methods to combine the measurements of validation datasets (others) that fall under the same
+    gridpoint of the dataset being validated (reference).
+
+    The goal is to include here all identified upscaling methods to provide an estimate at the reference footprint scale.
+
+    Implemented methods:
+
+        * time-stability filtering
+        * simple averaging
+
+    Parameters
+    ----------
+    ref_class: pygeogrids Grid obj.
+        Class containing the method read_ts for reading the data of the reference
+    others_class: dict
+        Dict of shape {'other_name': <pygeogrids Grid object>} for the other dataset
+    geo_subset: tuple, optional. Default is None.
+        Information on the geographic subset for data averaging -> (latmin, latmax, lonmin, lonmax)
+    manager_parms: dict
+        Dict of DataManager attributes
+    """
+
+    def __init__(
+            self,
+            ref_class,
+            others_class,
+            geo_subset,
+            manager_parms,
+    ):
+        self.ref_class = ref_class
+        self.others_class = others_class
+        self.geo_subset = geo_subset
+        # attributes used by the Mixin class:
+        self.datasets = manager_parms["datasets"]
+        self.period = manager_parms["period"]
+        self.read_ts_names = manager_parms["read_ts_names"]
+
+    @property
+    def lut(self) -> dict:
+        """Get a lookup table that combines the points falling under the same reference pixel"""
+        lut = {}
+        for other_name, other_class in self.others_class.items():
+            if hasattr(other_class, "grid"):  # todo: check potential object attributes
+                grid = other_class.grid
+            else:
+                grid = other_class.cls.cls.grid
```

Comment:
I think we should fix the adapters instead of relying on this approach. You can leave it for now, but let's add a task for this sprint.

Reply:
I actually already fixed it. If you rebase/merge the current master, `grid = other_class.grid` should also work with adapters (but the grid could be None in case the original reader had no grid).